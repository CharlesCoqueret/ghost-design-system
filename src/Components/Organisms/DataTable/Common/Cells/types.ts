import {
  IColumnType,
  IExtraEditableDataTableProps,
  IExtraLineEditableInPlaceDataTableProps,
  IExtraStaticDataTableProps,
} from '../types';

export interface ICellProps<T, U extends IColumnType<T>> {
  column: U;
  dataTestId?: string;
  editing?: boolean;
  extra?: IExtraStaticDataTableProps<T> | IExtraLineEditableInPlaceDataTableProps<T> | IExtraEditableDataTableProps<T>;
  forcedValue?: T[keyof T] | number | string | Date | null;
  onChange?: (newValue: T[keyof T]) => void;
  row?: T;
  rowIndex: number;
}
