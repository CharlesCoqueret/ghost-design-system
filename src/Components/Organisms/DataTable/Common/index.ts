export { ColumnType, SortDirectionEnum } from './types';
export type {
  IButtonCellProps,
  IColumnType,
  IExtraEditableDataTableProps,
  IExtraLineEditableInPlaceDataTableProps,
  IExtraStaticDataTableProps,
  TableType,
} from './types';
