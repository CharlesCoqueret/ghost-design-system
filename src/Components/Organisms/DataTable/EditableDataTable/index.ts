export type { IEditableDataTableProps } from './EditableDataTable';
export { default as EditableDataTable } from './EditableDataTable';
