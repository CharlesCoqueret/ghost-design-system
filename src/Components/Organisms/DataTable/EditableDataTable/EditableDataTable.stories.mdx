import { Canvas, Meta, ArgsTable, Story, Source } from '@storybook/addon-docs';
import LinkTo from '@storybook/addon-links/react';

import EditableDataTable from './EditableDataTable';
import { ColumnType } from '../Common/types';
import FakeExtraEditableDataTable from '../Common/FakeComponents/FakeExtraEditableDataTable';
import { Icon } from '../../../Atoms';

export const TemplateEditableDataTable = (args) => {
  return <EditableDataTable {...args} />;
};

<Meta
  title='Organism/DataTable'
  component={EditableDataTable}
  parameters={{
    viewMode: 'docs',
    layout: 'fullscreen',
  }}
/>

# EditableDataTable

- The editable data table displays information in a grid-like format of rows and columns. It organizes
  information in an easy to read and scan through way.
- Edition is always active when enabled (column and row), and is happening in place, making it less
  convenient for guidance, and error handling.
- If data validation is required, please use <LinkTo story="Line-Editable-Data-Table">LineEditableDataTable</LinkTo>.

Play with EditableDataTable live on [CodeSandbox](https://codesandbox.io/s/editabledatatable-demo-2tbec3)

## Editable column

Using the `editable` property of a column, you can enable edition of a column.

<Canvas>
  <EditableDataTable
    columns={[
      { suffix: '€', dataIndex: 'amount', editable: true, title: 'Amount (editable)', type: ColumnType.AMOUNT },
      { dataIndex: 'number', editable: false, title: 'Number (non-editable)', type: ColumnType.NUMBER },
      { dataIndex: 'percentage', editable: true, title: 'Percentage (editable)', type: ColumnType.PERCENTAGE },
    ]}
    data={[
      { amount: 0, number: 0, percentage: 0 },
      { amount: 1, number: 10, percentage: 100 },
      { amount: 2, number: 20, percentage: 200 },
    ]}
    extra={{
      onEdit: (row, dataIndex, editedRowIndex) => {
        console.log(
          `Edited row: ${JSON.stringify(row)}, edited data index: ${dataIndex}, edited row index: ${editedRowIndex}`,
        );
      },
    }}
  />
</Canvas>

## Editable row

Using `isEditable` and `onEdit` you can enable edition and manage the edition of a row.

<Canvas>
  <EditableDataTable
    columns={[
      { dataIndex: 'text', editable: true, title: 'Text (editable)', type: ColumnType.TEXT },
      { dataIndex: 'number', editable: false, title: 'Number (non-editable)', type: ColumnType.NUMBER },
      { dataIndex: 'percentage', editable: true, title: 'Percentage (editable)', type: ColumnType.PERCENTAGE },
    ]}
    data={[
      { text: 'Editable row 0', number: 0, percentage: 0 },
      { text: 'Non-editable row 1', number: 10, percentage: 100 },
      { text: 'Editable row 2', number: 20, percentage: 200 },
    ]}
    extra={{
      onEdit: (row, dataIndex, editedRowIndex) => {
        console.log(
          `Edited row: ${JSON.stringify(row)}, edited data index: ${dataIndex}, edited row index: ${editedRowIndex}`,
        );
      },
      isEditable: (row, rowIndex) => !row.text.includes('Non-editable'),
    }}
  />
</Canvas>

## Downloadable row

Using `isDownloadable` and `onRowDownload` you can enable download and manage the download of a row.

<Canvas>
  <EditableDataTable
    columns={[
      { dataIndex: 'text', editable: true, title: 'Text', type: ColumnType.TEXT },
      { dataIndex: 'number', editable: true, title: 'Number', type: ColumnType.NUMBER },
      { dataIndex: 'percentage', editable: true, title: 'Percentage', type: ColumnType.PERCENTAGE },
    ]}
    data={[
      { text: 'Downloadable row 0', number: 0, percentage: 0 },
      { text: 'Non-downloadable row 1', number: 10, percentage: 100 },
      { text: 'Downloadable row 2', number: 20, percentage: 200 },
    ]}
    extra={{
      onEdit: (row, dataIndex, editedRowIndex) => {
        console.log(
          `Edited row: ${JSON.stringify(row)}, edited data index: ${dataIndex}, edited row index: ${editedRowIndex}`,
        );
      },
      isDownloadable: (row, rowIndex) => !row.text.includes('Non-downloadable'),
      onRowDownload: (row, rowIndex) => {
        console.log(`Download requested for row: ${JSON.stringify(row)}, at row index: ${rowIndex}`);
      },
      localization: {
        actionColumn: 'Custom actions',
        downloadButton: 'Custom download',
      },
    }}
  />
</Canvas>

## Deletable row

Using `isDeletable` and `onRowDelete` you can enable deletion and manage the deletion of a row.

<Canvas>
  <EditableDataTable
    columns={[
      { dataIndex: 'text', editable: true, title: 'Text', type: ColumnType.TEXT },
      { dataIndex: 'number', editable: true, title: 'Number', type: ColumnType.NUMBER },
      { dataIndex: 'percentage', editable: true, title: 'Percentage', type: ColumnType.PERCENTAGE },
    ]}
    data={[
      { text: 'Deletable row 0', number: 0, percentage: 0 },
      { text: 'Non-deletable row 1', number: 10, percentage: 100 },
      { text: 'Deletable row 2', number: 20, percentage: 200 },
    ]}
    extra={{
      onEdit: (row, dataIndex, editedRowIndex) => {
        console.log(
          `Edited row: ${JSON.stringify(row)}, edited data index: ${dataIndex}, edited row index: ${editedRowIndex}`,
        );
      },
      isDeletable: (row, rowIndex) => !row.text.includes('Non-deletable'),
      onRowDelete: (row, rowIndex) => {
        console.log(`Delete requested for row: ${JSON.stringify(row)}, at row index: ${rowIndex}`);
      },
      localization: {
        deleteButton: 'Custom Delete',
        deletePopoverMessage: 'Custom do you confirm deletion ?',
        deletePopoverConfirm: 'Custom comfirm',
        deletePopoverCancel: 'Custom cancel',
      },
    }}
  />
</Canvas>

## Add new row

Using `canAddNewLine` and `onNewLine` you can enable new row creation and manage the initial value of a new row.

<Canvas>
  <EditableDataTable
    columns={[
      { dataIndex: 'text', editable: false, title: 'Text', type: ColumnType.TEXT },
      { dataIndex: 'number', editable: true, title: 'Number', type: ColumnType.NUMBER },
      { dataIndex: 'percentage', editable: true, title: 'Percentage', type: ColumnType.PERCENTAGE },
    ]}
    data={[
      { text: 'Existing row 0', number: 0, percentage: 0 },
      { text: 'Existing row 1', number: 10, percentage: 100 },
      { text: 'Existing row 2', number: 20, percentage: 200 },
    ]}
    extra={{
      onEdit: (row, dataIndex, editedRowIndex) => {
        console.log(
          `Edited row: ${JSON.stringify(row)}, edited data index: ${dataIndex}, edited row index: ${editedRowIndex}`,
        );
      },
      canAddNewLine: () => true,
      onNewLine: () => {
        const initialRow = { text: 'Newly created row', number: 100, percentage: 10 };
        console.log(`Returning the default object: ${JSON.stringify(initialRow)}`);
        return initialRow;
      },
    }}
  />
</Canvas>

## Props

<ArgsTable of={EditableDataTable} sort='requiredFirst' exclude={['dataTestId']} />

## extra: IExtraEditableDataTableProps<T\>

<ArgsTable of={FakeExtraEditableDataTable} sort='requiredFirst' />

## Story

<Canvas>
  <Story
    name='EditableDataTable'
    args={{
      data: [
        {
          name: 'Lorem ipsum dolor sit amet consectetur adipiscing elit',
          status: 'INACTIVE',
          price: -100000,
          parts: -10,
          startDate: new Date(2021, 2, 24),
        },
        {
          name: 'Consecte',
          status: 'ACTIVE',
          price: -2500000,
          parts: 15,
          startDate: new Date(2021, 1, 13),
        },
        {
          name: 'Sed',
          status: 'IN_PROGRESS',
          price: 150000,
          parts: 3,
          startDate: new Date(2021, 1, 4),
        },
        {
          name: 'dolore eu fugiat nulla pariatur',
          status: 'IN_PROGRESS',
          price: -354000,
          parts: 70,
          startDate: new Date(2021, 1, 15),
        },
        {
          name: 'Sed ut perspiciatis',
          status: 'ACTIVE',
          price: 4150000,
          parts: 30,
          startDate: new Date(2021, 2, 17),
        },
        {
          name: 'Ut enim ad minima veniam',
          status: 'ACTIVE',
          price: -70500,
          parts: 10,
          startDate: new Date(2021, 3, 10),
        },
      ],
      columns: [
        {
          title: 'Text',
          dataIndex: 'name',
          editable: true,
          type: ColumnType.TEXT,
          ellipsis: true,
          sorter: true,
          width: '20%',
        },
        {
          title: 'Badge',
          dataIndex: 'status',
          editable: true,
          sorter: true,
          type: ColumnType.BADGE,
          options: [
            { label: 'Inactive', value: 'INACTIVE' },
            { label: 'In progress', value: 'IN_PROGRESS' },
            { label: 'Active', value: 'ACTIVE' },
          ],
        },
        {
          title: 'Amount',
          dataIndex: 'price',
          editable: true,
          sorter: true,
          type: ColumnType.AMOUNT,
          suffix: '€',
        },
        {
          title: 'Percentage',
          dataIndex: 'parts',
          editable: true,
          sorter: true,
          type: ColumnType.PERCENTAGE,
        },
        {
          title: 'Date',
          dataIndex: 'startDate',
          editable: true,
          sorter: true,
          type: ColumnType.DATE,
        },
      ],
      extra: {
        computeTotal: (data, dataIndex) => {
          if (dataIndex === 'price' || dataIndex === 'parts') {
            return data
              .map((row) => {
                return row[dataIndex];
              })
              .reduce((a, b) => Number(a) + Number(b), 0);
          } else {
            return '-';
          }
        },
        onEdit: (row, dataIndex, rowIndex) => {
          console.log(`edited row: ${JSON.stringify(row)}, data index: ${dataIndex}, row index: ${rowIndex}`);
        },
        onRowClick: (row) => {
          console.log(`clicked row: ${JSON.stringify(row)}`);
        },
        onRowSelect: (rows, row) => {
          console.log(`Number of rows selected: ${rows.length}\nSelected row: ${row.name}`);
        },
        canAddNewLine: () => true,
        onNewLine: () => {
          const initialRow = { text: 'Newly created row', number: 100, percentage: 10 };
          console.log(`Returning the default object: ${JSON.stringify(initialRow)}`);
          return initialRow;
        },
        onRowDelete: (row, rowIndex) => {
          console.log(`Delete requested for row: ${JSON.stringify(row)}, at row index: ${rowIndex}`);
        },
        onRowDownload: (row, rowIndex) => {
          console.log(`Download requested for row: ${JSON.stringify(row)}, at row index: ${rowIndex}`);
        },
      },
    }}>
    {TemplateEditableDataTable.bind({})}
  </Story>
</Canvas>
