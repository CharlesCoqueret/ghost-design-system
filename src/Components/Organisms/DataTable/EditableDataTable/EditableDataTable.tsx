import React, { ReactElement, useCallback, useEffect, useState } from 'react';

import Button, { ColorButtonEnum } from '../../../Molecules/Button/Button';
import StaticDataTableFooter from '../StaticDataTable/StaticDataTableFooter';
import StaticDataTableHeader from '../StaticDataTable/StaticDataTableHeader';
import { ColumnType, IColumnType, IExtraEditableDataTableProps, SortDirectionEnum } from '../Common/types';
import EditableDataTableBody from './EditableDataTableBody';

export interface IEditableDataTableProps<T> {
  columns: Array<IColumnType<T>>;
  data: Array<T>;
  extra: IExtraEditableDataTableProps<T>;
  loading?: ReactElement;
  onSortChange?: (sortField?: keyof T, sortDirection?: SortDirectionEnum) => void;
}

const EditableDataTable = <T,>(props: IEditableDataTableProps<T>): ReactElement => {
  const { data, columns, extra, loading, onSortChange } = props;

  const [currentData, setCurrentData] = useState<Array<T>>(data);
  const [sortField, setSortField] = useState<keyof T | undefined>();
  const [sortDirection, setSortDirection] = useState<SortDirectionEnum | undefined>();

  const currentColumns: Array<IColumnType<T>> =
    !extra?.onRowDelete && !extra?.onRowDownload
      ? columns
      : [
          ...columns.filter((column) => column.type !== ColumnType.BUTTON),
          {
            title: extra?.localization?.actionColumn ?? 'Actions',
            type: ColumnType.BUTTON,
            moreActionsMessage: extra?.localization?.moreActionsMessage ?? 'More actions',
            buttons: [
              {
                hidden: (row, rowIndex) => {
                  if (!extra?.onRowDelete) return true;
                  if (extra.isDeletable === undefined || extra.isDeletable(row, rowIndex)) {
                    return false;
                  }
                  return true;
                },
                icon: ['fal', 'trash-alt'],
                label: extra?.localization?.deleteButton ?? 'Delete',
                onClick: (row, rowIndex) => {
                  if (extra?.onRowDelete) {
                    extra.onRowDelete(row, rowIndex);
                  }
                  setCurrentData((prev) => [...prev.filter((_item, index) => index !== rowIndex)]);
                },
                popover: {
                  message: extra?.localization?.deletePopoverMessage ?? 'Delete?',
                  cancel: extra?.localization?.deletePopoverCancel ?? 'Cancel',
                  confirm: extra?.localization?.deletePopoverConfirm ?? 'Confirm',
                },
                dataTestId: 'delete',
              },
              {
                hidden: (row, rowIndex) => {
                  if (!extra?.onRowDownload) return true;
                  if (extra.isDownloadable === undefined || extra.isDownloadable(row, rowIndex)) {
                    return false;
                  }
                  return true;
                },
                icon: ['fal', 'arrow-to-bottom'],
                label: extra?.localization?.downloadButton ?? 'Download',
                onClick: (row, rowIndex) => {
                  if (extra?.onRowDownload) {
                    extra.onRowDownload(row, rowIndex);
                  }
                },
                dataTestId: 'download',
              },
            ],
          },
        ];

  // Updating local copy of data whenever the provided data changes.
  useEffect(() => {
    setCurrentData(data);
  }, [data]);

  const handleSortChange = useCallback((newSortField: keyof T, newSortDirection?: SortDirectionEnum) => {
    if (sortField !== newSortField || newSortDirection !== newSortDirection) {
      setSortField(newSortField);
      setSortDirection(newSortDirection);

      if (onSortChange) {
        if (newSortField && newSortDirection) onSortChange(newSortField, newSortDirection);
        else onSortChange();
      }
    }
  }, []);

  const handleUpdateDataChange = (rowIndex: number, dataIndex: keyof T, newData: T[keyof T]) => {
    setCurrentData((prev) => {
      prev[rowIndex][dataIndex] = newData;
      return [...prev];
    });
    extra.onEdit({ ...currentData[rowIndex], [dataIndex]: newData }, dataIndex, rowIndex);
  };

  const addNewLine = () => {
    if (extra?.onNewLine === undefined) {
      console.warn('Missing onNewLine function');
      return;
    }
    const newLine = extra.onNewLine();
    setCurrentData((prev) => {
      prev.push(newLine);
      return [...prev];
    });
  };

  return (
    <>
      <table className='gds-table'>
        <StaticDataTableHeader<T>
          columns={currentColumns}
          onSortChange={handleSortChange}
          sortField={sortField}
          sortDirection={sortDirection}
          extra={extra}
        />
        <EditableDataTableBody<T>
          columns={currentColumns}
          data={currentData}
          extra={extra}
          handleUpdateDataChange={handleUpdateDataChange}
          loading={loading}
        />
        <StaticDataTableFooter<T> columns={currentColumns} data={currentData} extra={extra} />
      </table>
      {extra?.canAddNewLine && extra?.canAddNewLine() && (
        <Button
          className='gds-table-new-line'
          color={ColorButtonEnum.PRIMARY}
          label={extra?.localization?.addRow ?? 'Add Line'}
          onClick={addNewLine}
        />
      )}
    </>
  );
};

export default EditableDataTable;
