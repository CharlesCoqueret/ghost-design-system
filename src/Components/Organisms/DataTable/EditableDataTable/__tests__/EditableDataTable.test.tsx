import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import EditableDataTable from '../EditableDataTable';
import { ColumnType, SortDirectionEnum } from '../../Common/types';

describe('EditableDataTable component', () => {
  it('EditableDataTable renders', () => {
    const onEditMock = jest.fn();
    const onSortChangeMock = jest.fn();

    const { container } = render(
      <EditableDataTable<{ number: number }>
        columns={[{ dataIndex: 'number', title: 'Number', type: ColumnType.NUMBER }]}
        data={[{ number: 1 }]}
        extra={{ onEdit: onEditMock }}
        onSortChange={onSortChangeMock}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it('EditableDataTable handles sort', () => {
    const onEditMock = jest.fn();
    const onSortChangeMock = jest.fn();

    const { container } = render(
      <EditableDataTable<{ number: number }>
        columns={[{ dataIndex: 'number', sorter: true, title: 'Number', type: ColumnType.NUMBER }]}
        data={[{ number: 1 }]}
        extra={{ onEdit: onEditMock }}
        onSortChange={onSortChangeMock}
      />,
    );

    expect(container).toMatchSnapshot();

    userEvent.tab();
    userEvent.keyboard('{Enter}');

    expect(onSortChangeMock).toBeCalledTimes(1);
    expect(onSortChangeMock).toBeCalledWith('number', SortDirectionEnum.DESC);

    userEvent.keyboard('{Enter}');

    expect(onSortChangeMock).toBeCalledTimes(2);
    expect(onSortChangeMock).toBeCalledWith('number', SortDirectionEnum.ASC);

    userEvent.keyboard('{Enter}');

    expect(onSortChangeMock).toBeCalledTimes(3);
    expect(onSortChangeMock).toBeCalledWith();
  });

  it('EditableDataTable handles change, onRowDelete and onRowDownload', () => {
    const isDeletableMock = jest.fn().mockImplementation(() => true);
    const isDownloadableMock = jest.fn().mockImplementation(() => true);
    const onEditMock = jest.fn();
    const onRowDeleteMock = jest.fn();
    const onRowDownloadMock = jest.fn();

    const { container } = render(
      <EditableDataTable<{ number: number }>
        columns={[{ dataIndex: 'number', editable: true, title: 'Number', type: ColumnType.NUMBER }]}
        data={[{ number: 1 }]}
        dataTestId='DATA-TEST-ID'
        extra={{
          isDeletable: isDeletableMock,
          isDownloadable: isDownloadableMock,
          onEdit: onEditMock,
          onRowDelete: onRowDeleteMock,
          onRowDownload: onRowDownloadMock,
        }}
      />,
    );

    expect(container).toMatchSnapshot();

    userEvent.tab();
    if (document.activeElement) {
      userEvent.clear(document.activeElement);
    }

    expect(onEditMock).toBeCalledTimes(1);
    expect(onEditMock).toBeCalledWith({ number: undefined }, 'number', 0);

    // Download
    const downloadButton = screen.getByTestId('DATA-TEST-ID-download');
    userEvent.click(downloadButton);

    expect(onRowDownloadMock).toBeCalledTimes(1);
    expect(onRowDownloadMock).toBeCalledWith({ number: undefined }, 0);

    // Delete and confirm in popover
    const deleteButton = screen.getByTestId('DATA-TEST-ID-delete');
    userEvent.click(deleteButton);

    userEvent.tab();
    userEvent.tab();
    userEvent.keyboard('{Enter}');

    expect(onRowDeleteMock).toBeCalledTimes(1);
    expect(onRowDeleteMock).toBeCalledWith({ number: undefined }, 0);
  });

  it('EditableDataTable handles onRowDelete and onRowDownload with custom localization', () => {
    const isDeletableMock = jest.fn().mockImplementation(() => false);
    const isDownloadableMock = jest.fn().mockImplementation(() => false);
    const onEditMock = jest.fn();
    const onRowDeleteMock = jest.fn();
    const onRowDownloadMock = jest.fn();
    const canAddNewLineMock = jest.fn().mockImplementation(() => true);

    const { container } = render(
      <EditableDataTable<{ number: number }>
        columns={[{ dataIndex: 'number', editable: true, title: 'Number', type: ColumnType.NUMBER }]}
        data={[{ number: 1 }]}
        extra={{
          canAddNewLine: canAddNewLineMock,
          isDeletable: isDeletableMock,
          isDownloadable: isDownloadableMock,
          localization: {
            actionColumn: 'actionColumn',
            addRow: 'addRow',
            deleteButton: 'deleteButton',
            deletePopoverCancel: 'deletePopoverCancel',
            deletePopoverConfirm: 'deletePopoverConfirm',
            deletePopoverMessage: 'deletePopoverMessage',
            downloadButton: 'downloadButton',
            moreActionsMessage: 'moreActionsMessage',
          },
          onEdit: onEditMock,
          onRowDelete: onRowDeleteMock,
          onRowDownload: onRowDownloadMock,
        }}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it('EditableDataTable handles canAddNewLine with onNewLine', () => {
    const canAddNewLineMock = jest.fn().mockImplementation(() => true);
    const onEditMock = jest.fn();
    const onNewLineMock = jest.fn();

    const { container } = render(
      <EditableDataTable<{ number: number }>
        columns={[{ dataIndex: 'number', title: 'Number', type: ColumnType.NUMBER }]}
        data={[{ number: 1 }]}
        extra={{
          canAddNewLine: canAddNewLineMock,
          onEdit: onEditMock,
          onNewLine: onNewLineMock,
        }}
      />,
    );

    expect(canAddNewLineMock).toBeCalledTimes(1);

    expect(container).toMatchSnapshot();

    userEvent.tab();
    userEvent.keyboard('{Enter}');

    expect(onNewLineMock).toBeCalledTimes(1);
  });

  it('EditableDataTable handles canAddNewLine without onNewLine', () => {
    console.error = jest.fn();
    const canAddNewLineMock = jest.fn().mockImplementation(() => true);
    const onEditMock = jest.fn();

    const { container } = render(
      <EditableDataTable<{ number: number }>
        columns={[{ dataIndex: 'number', title: 'Number', type: ColumnType.NUMBER }]}
        data={[{ number: 1 }]}
        extra={{
          canAddNewLine: canAddNewLineMock,
          onEdit: onEditMock,
        }}
      />,
    );

    expect(canAddNewLineMock).toBeCalledTimes(1);

    expect(container).toMatchSnapshot();

    userEvent.tab();
    userEvent.keyboard('{Enter}');

    expect(console.error).toBeCalledTimes(1);
    expect(console.error).toBeCalledWith('Missing onNewLine function');
  });
});
