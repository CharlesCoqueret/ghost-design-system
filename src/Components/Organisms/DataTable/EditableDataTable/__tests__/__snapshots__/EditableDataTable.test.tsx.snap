// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditableDataTable component EditableDataTable handles canAddNewLine with onNewLine 1`] = `
<div>
  <div>
    <table
      class="gds-table"
    >
      <thead>
        <tr>
          <th
            class=""
          >
            <div
              class="table--header-value"
              tabindex="-1"
            >
              <span
                class="typography table--header-value--title right ellipsis"
                title="Number"
              >
                Number
              </span>
            </div>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          class=""
          tabindex="-1"
        >
          <td
            class=""
          >
            <div
              class="fieldGroup"
            >
              <span
                autocomplete="off"
                class="default readOnly table--cell-value--number-readonly"
                id="number"
                name="number"
              >
                1
              </span>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
    <button
      class="button gds-table-new-line"
      color="primary"
      tabindex="0"
      type="button"
    >
      <div
        class="label"
      >
        Add row
      </div>
    </button>
  </div>
</div>
`;

exports[`EditableDataTable component EditableDataTable handles canAddNewLine without onNewLine 1`] = `
<div>
  <div>
    <table
      class="gds-table"
    >
      <thead>
        <tr>
          <th
            class=""
          >
            <div
              class="table--header-value"
              tabindex="-1"
            >
              <span
                class="typography table--header-value--title right ellipsis"
                title="Number"
              >
                Number
              </span>
            </div>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          class=""
          tabindex="-1"
        >
          <td
            class=""
          >
            <div
              class="fieldGroup"
            >
              <span
                autocomplete="off"
                class="default readOnly table--cell-value--number-readonly"
                id="number"
                name="number"
              >
                1
              </span>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
    <button
      class="button gds-table-new-line"
      color="primary"
      tabindex="0"
      type="button"
    >
      <div
        class="label"
      >
        Add row
      </div>
    </button>
  </div>
</div>
`;

exports[`EditableDataTable component EditableDataTable handles change, onRowDelete and onRowDownload 1`] = `
<div>
  <div>
    <table
      class="gds-table"
    >
      <thead>
        <tr>
          <th
            class=""
          >
            <div
              class="table--header-value"
              tabindex="-1"
            >
              <span
                class="typography table--header-value--title right ellipsis"
                title="Number"
              >
                Number
              </span>
            </div>
          </th>
          <th
            class=""
          >
            <div
              class="table--header-value"
              tabindex="-1"
            >
              <span
                class="typography table--header-value--title center ellipsis"
                title="Actions"
              >
                Actions
              </span>
            </div>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          class=""
          tabindex="-1"
        >
          <td
            class=""
          >
            <div
              class="fieldGroup"
            >
              <input
                autocomplete="off"
                class="default"
                data-testid="DATA-TEST-ID-Number-0"
                inputmode="numeric"
                name="number"
                type="text"
                value="1"
              />
            </div>
          </td>
          <td
            class=""
          >
            <div
              class="table--cell--value--button"
            >
              <button
                class="button"
                color="reversed"
                data-testid="DATA-TEST-ID-delete"
                tabindex="0"
                type="button"
              >
                <div
                  class="icon"
                >
                  <i
                    class="icon"
                    icon="fal,trash-alt"
                  />
                </div>
              </button>
              <div>
                Delete
              </div>
              <div>
                <div>
                  ControlleMenu
                  closed
                  <div
                    class="container"
                  >
                    <div
                      class="title"
                    >
                      Delete?
                    </div>
                    <div
                      class="buttons"
                    >
                      <button
                        class="button"
                        color="secondary"
                        data-testid="cancel"
                        tabindex="0"
                        type="button"
                      >
                        <div
                          class="label"
                        >
                          Cancel
                        </div>
                      </button>
                      <button
                        class="button"
                        color="primary"
                        data-testid="confirm"
                        tabindex="0"
                        type="button"
                      >
                        <div
                          class="label"
                        >
                          Confirm
                        </div>
                      </button>
                    </div>
                  </div>
                  has ref
                </div>
              </div>
              <button
                class="button"
                color="reversed"
                data-testid="DATA-TEST-ID-download"
                tabindex="0"
                type="button"
              >
                <div
                  class="icon"
                >
                  <i
                    class="icon"
                    icon="fal,arrow-down-to-line"
                  />
                </div>
              </button>
              <div>
                Download
              </div>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`EditableDataTable component EditableDataTable handles onRowDelete and onRowDownload with custom localization 1`] = `
<div>
  <div>
    <table
      class="gds-table"
    >
      <thead>
        <tr>
          <th
            class=""
          >
            <div
              class="table--header-value"
              tabindex="-1"
            >
              <span
                class="typography table--header-value--title right ellipsis"
                title="Number"
              >
                Number
              </span>
            </div>
          </th>
          <th
            class=""
          >
            <div
              class="table--header-value"
              tabindex="-1"
            >
              <span
                class="typography table--header-value--title center ellipsis"
                title="actionColumn"
              >
                actionColumn
              </span>
            </div>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          class=""
          tabindex="-1"
        >
          <td
            class=""
          >
            <div
              class="fieldGroup"
            >
              <input
                autocomplete="off"
                class="default"
                inputmode="numeric"
                name="number"
                type="text"
                value="1"
              />
            </div>
          </td>
          <td
            class=""
          >
            <div
              class="table--cell--value--button"
            />
          </td>
        </tr>
      </tbody>
    </table>
    <button
      class="button gds-table-new-line"
      color="primary"
      tabindex="0"
      type="button"
    >
      <div
        class="label"
      >
        addRow
      </div>
    </button>
  </div>
</div>
`;

exports[`EditableDataTable component EditableDataTable handles sort 1`] = `
<div>
  <div>
    <table
      class="gds-table"
    >
      <thead>
        <tr>
          <th
            class=""
          >
            <div
              class="table--header-value table--header-value--sortable"
              tabindex="0"
            >
              <span
                class="typography table--header-value--title right ellipsis"
                title="Number"
              >
                Number
              </span>
              <span
                class="table--header-value--sorter"
              >
                <i
                  class="icon"
                  icon="fas,sort"
                />
              </span>
            </div>
            <div>
              Click to sort
            </div>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          class=""
          tabindex="-1"
        >
          <td
            class=""
          >
            <div
              class="fieldGroup"
            >
              <span
                autocomplete="off"
                class="default readOnly table--cell-value--number-readonly"
                id="number"
                name="number"
              >
                1
              </span>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`EditableDataTable component EditableDataTable renders 1`] = `
<div>
  <div>
    <table
      class="gds-table"
    >
      <thead>
        <tr>
          <th
            class=""
          >
            <div
              class="table--header-value"
              tabindex="-1"
            >
              <span
                class="typography table--header-value--title right ellipsis"
                title="Number"
              >
                Number
              </span>
            </div>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          class=""
          tabindex="-1"
        >
          <td
            class=""
          >
            <div
              class="fieldGroup"
            >
              <span
                autocomplete="off"
                class="default readOnly table--cell-value--number-readonly"
                id="number"
                name="number"
              >
                1
              </span>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
